src/Chainweb/Rosetta/Internal.hs

```diff
@@ -351,66 +383,106 @@ nonGenesisTransaction logs initial rest target
+--       then assumes that this log corresponds to the given remediation.
+--     Otherwise, a rosetta transaction is created for the given
+--       remediation with an empty operations list.
+-- NOTE: Assumes that each remediation must have been successful and thus
+--       would have a TxId associated with it.
+-- NOTE: Not all remediations transactions will output coin table logs.
+remediationAcc
+    :: AccumulatorType (TxAccumulator rosettaTx)
+    -> TxAccumulator rosettaTx
+    -> (Command payload, TxId)
+    -> TxAccumulator rosettaTx
+remediationAcc accTyp acc (remTx, remTid) =
+  case _txAccumulator_logsLeft acc of
+    (logTid,logs):rest
+      | logTid == remTid -> -- remediation touched coin table
+        let ops = indexedOperations $!
+              UnindexedOperations
+              { _unindexedOperation_fundOps = []
+              , _unindexedOperation_transferOps = makeOps (logTid, logs)
+              , _unindexedOperation_gasOps = []
+              }
+            rosettaTx = rosettaTransactionFromCmd remTx $! ops
+        in makeAcc rest rosettaTx
+    rest -> -- list of logs empty or remediation didn't touch coin table
+      makeAcc rest $!
+      rosettaTransactionFromCmd remTx []
+  where
+    makeAcc restLogs rosettaTx =
+      accumulatorFunction accTyp acc restLogs rosettaTx
+    makeOps (tid, logs) =
+      map (operation Remediation TransferOrCreateAcct tid) logs
+
+-- | Matches all transactions in a remediation block
+--   (including coinbase, remediations, and user transactions)
+--   to their coin table logs.
+--  Coinbase logs are matched first, followed by remediations txs,
+--  and finally user transactions are matched using the same algorithm
+--  as non-genesis transactions.
+-- NOTE: Assumes that each remediation transaction incremented
+--       the TxId counter by one.
+-- NOTE: Assumes remediations don't occur in blocks where coinbase transaction
+--       could have failed. It needs to know the TxId of the coinbase transaction in
+--       order to derive the remediation's TxIds.
 remediations
     :: Map TxId [AccountLog]
+    -> ChainId
     -> CoinbaseTx (CommandResult Hash)
-    -> Either RosettaFailure [Transaction]
-remediations logs initial =
-    overwriteError RosettaMismatchTxLogs work
+    -- Remediation transactions.
+    -- NOTE: No CommandResult available for these.
+    -> [Command payload]
+    -- User transactions in the same block as remediations
+    -> V.Vector (CommandResult Hash)
+    -> Either String [Transaction]
+remediations logs cid coinbase remTxs txs = do
+  TxAccumulator restLogs coinbaseTx <- nonGenesisCoinbaseLog logsList cid coinbase
+  coinbaseTxId <- note "remediations: No TxId found for Coinbase" (_crTxId coinbase)
+
+  let remWithTxIds = zip remTxs [(succ coinbaseTxId)..]
+      -- Assumes that each remediation transaction gets its own TxId.
+      -- Assumes that TxIds are going to be sequential for each remediation.
```

Comment:
This assumption holds, since the remediation txs are applied _directly after coinbase at a particular height_, as part of `applyCoinbase`. We construct the blocks, hence, the txids are not random.